### ISSUE TITLE
Performance tests broken?

### ISSUE THREAD

**Issue by jkinable** [jgrapht/jgrapht#1047]
Did someone recently run any of the performance tests? When running our `MaximalCliqueEnumerationPerformanceTest` from within IntelliJ I get:

```
/usr/lib/jvm/java-14-oracle/bin/java -ea -Didea.test.cyclic.buffer.size=1048576 -javaagent:/home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/lib/idea_rt.jar=39437:/home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/bin --patch-module org.jgrapht.core=/home/ANT.AMAZON.COM/kinable/workspace/research/external/jgrapht/jgrapht/jgrapht-core/target/test-classes --add-reads org.jgrapht.core=ALL-UNNAMED --add-opens org.jgrapht.core/org.jgrapht.perf.clique=ALL-UNNAMED --add-modules org.jgrapht.core -Dfile.encoding=UTF-8 -classpath /home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/lib/idea_rt.jar:/home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/plugins/junit/lib/junit5-rt.jar:/home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/plugins/junit/lib/junit-rt.jar:/home/ANT.AMAZON.COM/kinable/workspace/research/external/jgrapht/jgrapht/jgrapht-core/target/test-classes:/home/ANT.AMAZON.COM/kinable/.m2/repository/com/googlecode/junit-toolbox/junit-toolbox/2.4/junit-toolbox-2.4.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/commons-io/commons-io/2.4/commons-io-2.4.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/junit/junit/4.13.1/junit-4.13.1.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/hamcrest/hamcrest-library/1.3/hamcrest-library-1.3.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/mockito/mockito-core/1.9.5/mockito-core-1.9.5.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/objenesis/objenesis/1.0/objenesis-1.0.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/openjdk/jmh/jmh-core/1.23/jmh-core-1.23.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/net/sf/jopt-simple/jopt-simple/4.6/jopt-simple-4.6.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/apache/commons/commons-math3/3.2/commons-math3-3.2.jar:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/openjdk/jmh/jmh-generator-annprocess/1.23/jmh-generator-annprocess-1.23.jar -p /home/ANT.AMAZON.COM/kinable/.m2/repository/org/jheaps/jheaps/0.13/jheaps-0.13.jar:/home/ANT.AMAZON.COM/kinable/workspace/research/external/jgrapht/jgrapht/jgrapht-core/target/classes com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit4 org.jgrapht.perf.clique.MaximalCliqueEnumerationPerformanceTest
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.openjdk.jmh.util.Utils (file:/home/ANT.AMAZON.COM/kinable/.m2/repository/org/openjdk/jmh/jmh-core/1.23/jmh-core-1.23.jar) to field java.io.PrintStream.charOut
WARNING: Please consider reporting this to the maintainers of org.openjdk.jmh.util.Utils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
# JMH version: 1.23
# VM version: JDK 14.0.1, Java HotSpot(TM) 64-Bit Server VM, 14.0.1+7
# VM invoker: /usr/lib/jvm/java-14-oracle/bin/java
# VM options: -ea -Didea.test.cyclic.buffer.size=1048576 -javaagent:/home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/lib/idea_rt.jar=39437:/home/ANT.AMAZON.COM/kinable/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/203.7148.57/bin --patch-module=org.jgrapht.core=/home/ANT.AMAZON.COM/kinable/workspace/research/external/jgrapht/jgrapht/jgrapht-core/target/test-classes --add-reads=org.jgrapht.core=ALL-UNNAMED --add-opens=org.jgrapht.core/org.jgrapht.perf.clique=ALL-UNNAMED --add-modules=org.jgrapht.core -Dfile.encoding=UTF-8 --module-path=/home/ANT.AMAZON.COM/kinable/.m2/repository/org/jheaps/jheaps/0.13/jheaps-0.13.jar:/home/ANT.AMAZON.COM/kinable/workspace/research/external/jgrapht/jgrapht/jgrapht-core/target/classes
# Warmup: 5 iterations, single-shot each
# Measurement: 10 iterations, single-shot each
# Timeout: 10 min per iteration
# Threads: 1 thread
# Benchmark mode: Single shot invocation time
# Benchmark: org.jgrapht.perf.clique.MaximalCliqueEnumerationPerformanceTest.BronKerboschRandomGraphBenchmark.run

# Run progress: 0.00% complete, ETA 00:00:00
# Fork: 1 of 1
# Warmup Iteration   1: <failure>

java.lang.IllegalAccessException: class org.openjdk.jmh.runner.BenchmarkHandler$1 cannot access class org.jgrapht.perf.clique.generated.MaximalCliqueEnumerationPerformanceTest_BronKerboschRandomGraphBenchmark_run_jmhTest (in module org.jgrapht.core) because module org.jgrapht.core does not export org.jgrapht.perf.clique.generated to unnamed module @39a001e8
	at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:376)
	at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:647)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:490)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at org.openjdk.jmh.runner.BenchmarkHandler$1.initialValue(BenchmarkHandler.java:82)
	at org.openjdk.jmh.runner.BenchmarkHandler$1.initialValue(BenchmarkHandler.java:78)
	at java.base/java.lang.ThreadLocal.setInitialValue(ThreadLocal.java:195)
	at java.base/java.lang.ThreadLocal.get(ThreadLocal.java:172)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:452)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:437)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)


Benchmark had encountered error, and fail on error was requested

org.openjdk.jmh.runner.RunnerException: Benchmark caught the exception

	at org.openjdk.jmh.runner.Runner.runBenchmarks(Runner.java:585)
	at org.openjdk.jmh.runner.Runner.internalRun(Runner.java:320)
	at org.openjdk.jmh.runner.Runner.run(Runner.java:209)
	at org.jgrapht.core/org.jgrapht.perf.clique.MaximalCliqueEnumerationPerformanceTest.testMaximalCliqueRandomGraphBenchmark(MaximalCliqueEnumerationPerformanceTest.java:123)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: org.openjdk.jmh.runner.BenchmarkException: Benchmark error during the run
	at org.openjdk.jmh.runner.BenchmarkHandler.runIteration(BenchmarkHandler.java:428)
	at org.openjdk.jmh.runner.BaseRunner.runBenchmark(BaseRunner.java:261)
	at org.openjdk.jmh.runner.BaseRunner.runBenchmark(BaseRunner.java:233)
	at org.openjdk.jmh.runner.BaseRunner.doSingle(BaseRunner.java:138)
	at org.openjdk.jmh.runner.BaseRunner.runBenchmarksForked(BaseRunner.java:75)
	at org.openjdk.jmh.runner.ForkedRunner.run(ForkedRunner.java:72)
	at org.openjdk.jmh.runner.ForkedMain.main(ForkedMain.java:84)
	Suppressed: java.lang.IllegalAccessException: class org.openjdk.jmh.runner.BenchmarkHandler$1 cannot access class org.jgrapht.perf.clique.generated.MaximalCliqueEnumerationPerformanceTest_BronKerboschRandomGraphBenchmark_run_jmhTest (in module org.jgrapht.core) because module org.jgrapht.core does not export org.jgrapht.perf.clique.generated to unnamed module @39a001e8
		at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:376)
		at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:647)
		at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:490)
		at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
		at org.openjdk.jmh.runner.BenchmarkHandler$1.initialValue(BenchmarkHandler.java:82)
		at org.openjdk.jmh.runner.BenchmarkHandler$1.initialValue(BenchmarkHandler.java:78)
		at java.base/java.lang.ThreadLocal.setInitialValue(ThreadLocal.java:195)
		at java.base/java.lang.ThreadLocal.get(ThreadLocal.java:172)
		at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:452)
		at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:437)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
		at java.base/java.lang.Thread.run(Thread.java:832)
```

**Comment by jsichi** [comment]
I can reproduce this problem with `mvn -Dtest=PerformanceTestSuite test` so it's not an IntelliJ issue.

**Comment by jsichi** [comment]
I can also repro against the 1.5.0 release itself (1.4.0 is fine), so looks like this has been broken ever since we added modules.

(I guess we should add running the performance tests to the runbook for our release process.)

**Comment by besza** [comment]
None of the `*TestSuite`s are running that are using "com.googlecode.junittoolbox".

However, each of them has an alternative:
* IntegrationTestSuite -> `mvn verify`
* FastTestSuite -> `mvn test`
* PerformanceTestSuite -> `mvn test -Dtest=**/perf/**/* -pl jgrapht-core`

**Comment by jkinable** [comment]
@besza Is there a way I can run a specific performance test? E.g. one of the clique performance tests?

Also, is this 'issue' something we can (and should?) fix? Or does this work as expected. I'm afraid this is well beyond my knowledge of modules.

**Comment by besza** [comment]
Sure, you can narrow the regexp `mvn test -Dtest=**/perf/clique/**/* -pl jgrapht-core` or you can just name the test you want with `-Dtest=TreeDynamicConnectivityPerformanceTest`.

I'm also new to the module system but I found this to be a great resource: https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world.html .

I don't think this is worth _fixing_. Eventually, "com.googlecode.junittoolbox" should be dropped in favor JUnit 5.

**Comment by jkinable** [comment]
Just naming the test you want to perform does not seem to work. See below.

`mvn test -Dtest=TreeDynamicConnectivityPerformanceTest -pl jgrapht-core`

Result:
```
Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project jgrapht-core: No tests were executed!
```

`mvn test -Dtest=MaximalCliqueEnumerationPerformanceTest -pl jgrapht-core`

Result:
```
# Run progress: 0.00% complete, ETA 00:00:00
# Fork: 1 of 1
# Warmup Iteration   1: <failure>

java.lang.IllegalAccessException: class org.openjdk.jmh.runner.BenchmarkHandler$1 cannot access class org.jgrapht.perf.clique.generated.MaximalCliqueEnumerationPerformanceTest_BronKerboschRandomGraphBenchmark_run_jmhTest (in module org.jgrapht.core) because module org.jgrapht.core does not export org.jgrapht.perf.clique.generated to unnamed module @6a59ddde
	at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:376)
	at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:647)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:490)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at org.openjdk.jmh.runner.BenchmarkHandler$1.initialValue(BenchmarkHandler.java:82)
	at org.openjdk.jmh.runner.BenchmarkHandler$1.initialValue(BenchmarkHandler.java:78)
	at java.base/java.lang.ThreadLocal.setInitialValue(ThreadLocal.java:195)
	at java.base/java.lang.ThreadLocal.get(ThreadLocal.java:172)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:452)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:437)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)


Benchmark had encountered error, and fail on error was requested
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.096 s <<< FAILURE! - in org.jgrapht.perf.clique.MaximalCliqueEnumerationPerformanceTest
[ERROR] org.jgrapht.perf.clique.MaximalCliqueEnumerationPerformanceTest.testMaximalCliqueRandomGraphBenchmark  Time elapsed: 1.081 s  <<< ERROR!
org.openjdk.jmh.runner.RunnerException: Benchmark caught the exception
	at org.jgrapht.core@1.5.1-SNAPSHOT/org.jgrapht.perf.clique.MaximalCliqueEnumerationPerformanceTest.testMaximalCliqueRandomGraphBenchmark(MaximalCliqueEnumerationPerformanceTest.java:123)
Caused by: org.openjdk.jmh.runner.BenchmarkException: Benchmark error during the run
```

This however does work:
`mvn test -Dtest=**/perf/clique/**/* -pl jgrapht-core`

**Comment by besza** [comment]
1/ `TreeDynamicConnectivityPerformanceTest` does not have a method annotated with `@Test` so the surefire plugin does not pick it up.
2/ `mvn test -Dsurefire.useModulePath=false -Dtest=MaximalCliqueEnumerationPerformanceTest -pl jgrapht-core` works and similarly `mvn test -Dsurefire.useModulePath=false -Dtest=PerformanceTestSuite` seems to work as well.

**Comment by syoon2** [comment]
@jsichi It seems that #1170 is a duplicate of this issue - would you verify? If so, this issue could safely be closed as it is fixed with #1178.
